ksql-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java

```diff
@@ -784,52 +808,32 @@ public Node visitLogicalBinary(SqlBaseParser.LogicalBinaryContext context) {
 
   @Override
   public Node visitJoinRelation(SqlBaseParser.JoinRelationContext context) {
-    Relation left = (Relation) visit(context.left);
-    Relation right;
-
-    if (context.CROSS() != null) {
-      right = (Relation) visit(context.right);
-      return new Join(
-          getLocation(context),
-          Join.Type.CROSS,
-          left,
-          right,
-          Optional.<JoinCriteria>empty()
-      );
-    }
-
-    JoinCriteria criteria;
-    if (context.NATURAL() != null) {
-      right = (Relation) visit(context.right);
-      criteria = new NaturalJoin();
-    } else {
-      right = (Relation) visit(context.rightRelation);
-      if (context.joinCriteria().ON() != null) {
-        criteria = new JoinOn((Expression) visit(context.joinCriteria().booleanExpression()));
-      } else if (context.joinCriteria().USING() != null) {
-        List<String> columns = context.joinCriteria()
-            .identifier().stream()
-            .map(AstBuilder::getIdentifierText)
-            .collect(toList());
-
-        criteria = new JoinUsing(columns);
-      } else {
-        throw new IllegalArgumentException("Unsupported join criteria");
-      }
+    if (context.joinCriteria().ON() == null) {
+      throw new KsqlException("Invalid join criteria specified. KSQL only supports joining on "
+                              + "column values. For example `... left JOIN right on left.col = "
+                              + "right.col ...`. ");
     }
 
+    JoinCriteria criteria =
+        new JoinOn((Expression) visit(context.joinCriteria().booleanExpression()));
     Join.Type joinType;
     if (context.joinType().LEFT() != null) {
       joinType = Join.Type.LEFT;
-    } else if (context.joinType().RIGHT() != null) {
-      joinType = Join.Type.RIGHT;
-    } else if (context.joinType().FULL() != null) {
-      joinType = Join.Type.FULL;
+    } else if (context.joinType().OUTER() != null) {
+      joinType = Join.Type.OUTER;
     } else {
       joinType = Join.Type.INNER;
     }
 
-    return new Join(getLocation(context), joinType, left, right, Optional.of(criteria));
+    SpanExpression spanExpression = null;
+    if (context.joinWindow() != null && context.joinWindow().spanExpression() != null) {
+      spanExpression = (SpanExpression) visitSpanExpression(
+          context.joinWindow().spanExpression());
+    }
+    Relation left = (Relation) visit(context.left);
```

Comment:
Should we add another subclass of `Relation` for `Stream`? I believe these will be two instances of `Table` at the moment which is a bit confusing later on

Reply:
Hmm. In this context, the relation can either be `AliasedRelation`, which is a stream or table, or `JoinRelation` which is a join of two streams or tables. This is defined in the grammar. 

The sub relations of a join should always be an `AliasedRelation`. This is also enforced in the grammar. I updated the code to check that the actual type of `left` and `right` is `AliasedRelation`. 

Further, I synced with @hjafarpour , and the `Table` in the grammar is basically a data source: a `Table` at the grammar level could wind up being a `Stream`. So all `AliasedRelations` start as tables and then may become streams.

Reply:
ok, leave it as is for now. i still find it confusing though!